pkg/state/state.go

```diff
@@ -0,0 +1,134 @@
+package state
+
+import (
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/bruin-data/bruin/pkg/pipeline"
+)
+
+var version = "dev"
+
+const (
+	StatusPending Status = "pending"
+	StatusRunning Status = "running"
+	StatusSuccess Status = "success"
+	StatusFailed  Status = "failed"
+	StatusSkipped Status = "skipped"
+)
+
+type State struct {
+	sync.RWMutex
+	Parameters map[string]string            `json:"parameters"`
+	Metadata   Metadata                     `json:"metadata"`
+	State      map[string][]*SchedulerState `json:"state"`
+	Version    string                       `json:"version"`
+	TimeStamp  time.Time                    `json:"timestamp"`
+	RunID      string                       `json:"run_id"`
+}
+
+type Metadata struct {
+	Version string `json:"version"`
+	OS      string `json:"os"`
+}
+
+type SchedulerState struct {
+	Upstream  []pipeline.Upstream `json:"upstream"`
+	Status    Status              `json:"status"`
+	Error     string              `json:"error"`
+	StartTime time.Time           `json:"start_time"`
+	EndTime   time.Time           `json:"end_time"`
+	Name      string              `json:"name"`
+	ID        string              `json:"id"`
+}
+
+type Status string
+
+func NewState(runid string, parameters map[string]string) *State {
+	return &State{
+		Parameters: parameters,
+		Metadata: Metadata{
+			Version: version,
+			OS:      runtime.GOOS,
+		},
+		State:     map[string][]*SchedulerState{},
+		Version:   "1.0.0",
+		TimeStamp: time.Time{},
+		RunID:     runid,
+	}
+}
+
+func (s *State) InitState(foundPipelines []pipeline.Pipeline) error {
```

Comment:
I wouldn't have done it this way, this means you'll have to update the state step by step at every execution from the scheduler, which is duplicate work. the scheduler already stores this state, I would couple it with that, and just create a struct that can be saved from the scheduler.